pkg/config/keys.go

```diff
@@ -0,0 +1,143 @@
+package config
+
+import (
+	"crypto"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/base64"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"os"
+
+	libp2p_crypto "github.com/libp2p/go-libp2p/core/crypto"
+	"github.com/spf13/viper"
+
+	"github.com/bacalhau-project/bacalhau/pkg/config/types"
+	"github.com/bacalhau-project/bacalhau/pkg/util/closer"
+)
+
+// GetClientPublicKeyString returns a base64-encoding of the user's public ID key:
+// NOTE: must be called after InitConfig() or system will panic.
+func GetClientPublicKeyString() (string, error) {
+	userIDKey, err := loadUserIDKey()
+	if err != nil {
+		return "", err
+	}
+
+	return encodePublicKey(&userIDKey.PublicKey), nil
+}
+
+func GetClientPublicKey() (*rsa.PublicKey, error) {
+	privKey, err := loadUserIDKey()
+	if err != nil {
+		return nil, err
+	}
+	return &privKey.PublicKey, nil
+}
+
+func GetClientPrivateKey() (*rsa.PrivateKey, error) {
+	privKey, err := loadUserIDKey()
+	if err != nil {
+		return nil, err
+	}
+	return privKey, nil
+}
+
+func GetClientID() (string, error) {
+	return loadClientID()
+}
+
+// loadClientID loads a hash identifying a user based on their ID key.
+func loadClientID() (string, error) {
+	key, err := loadUserIDKey()
+	if err != nil {
+		return "", fmt.Errorf("failed to load user ID key: %w", err)
+	}
+
+	return convertToClientID(&key.PublicKey), nil
+}
+
+const (
+	sigHash = crypto.SHA256 // hash function to use for sign/verify
+)
+
+// convertToClientID converts a public key to a client ID:
+func convertToClientID(key *rsa.PublicKey) string {
+	hash := sigHash.New()
+	hash.Write(key.N.Bytes())
+	hashBytes := hash.Sum(nil)
+
+	return fmt.Sprintf("%x", hashBytes)
+}
+
+// encodePublicKey encodes a public key as a string:
+func encodePublicKey(key *rsa.PublicKey) string {
+	return base64.StdEncoding.EncodeToString(x509.MarshalPKCS1PublicKey(key))
+}
+
+// loadUserIDKey loads the user ID key from whatever source is configured.
+func loadUserIDKey() (*rsa.PrivateKey, error) {
+	keyFile := viper.GetString(types.UserUserKeyPath)
```

Comment:
This stutters because of the config structure:
```go
type BacalhauConfig struct {
	Node    NodeConfig
	User    UserConfig
	Metrics MetricsConfig
}

type UserConfig struct {
	UserKeyPath   string
	Libp2pKeyPath string
}
```
e.g. the path to the value in the config is `User.UserKeyPath`, we should probably just refer to it as `User.KeyPath`. What do you think?

Reply:
`User.KeyPath` is probably a bit more parseable.

Reply:
done in d4450bf0db5bdbeb013a10c843b6f6254a34022a